pytorch_lightning/overrides/data_parallel.py

```diff
@@ -150,73 +152,75 @@ def parallel_apply(self, replicas, inputs, kwargs):
+            " From now on we recommend to directly sublcass `torch.nn.parallel.DistributedDataParallel`.",
+            DeprecationWarning
+        )
+        super().__init__(LightningDistributedModule(module), *args, **kwargs)
 
-    def forward(self, *inputs, **kwargs):  # pragma: no-cover
-        self._sync_params()
-        self.reducer_reset_hooks()
-        fx_called: str = ''
-
-        if self.device_ids:
-
-            inputs, kwargs = self.scatter(inputs, kwargs, self.device_ids)
-            if len(self.device_ids) == 1:
-                # --------------
-                # LIGHTNING MOD
-                # --------------
-                # normal
-                # output = self.module(*inputs[0], **kwargs[0])
-                # lightning
-                if self.module.training:
-                    output = self.module.training_step(*inputs[0], **kwargs[0])
-                    fx_called = 'training_step'
-                elif self.module.testing:
-                    output = self.module.test_step(*inputs[0], **kwargs[0])
-                    fx_called = 'test_step'
-                else:
-                    output = self.module.validation_step(*inputs[0], **kwargs[0])
-                    fx_called = 'validation_step'
-            else:
-                outputs = self.parallel_apply(self._module_copies[:len(inputs)], inputs, kwargs)
-                output = self.gather(outputs, self.output_device)
-        else:
-            # output = self.module(*inputs, **kwargs)
-            # normal lightning (ddp_cpu)
-            if self.module.training:
-                output = self.module.training_step(*inputs, **kwargs)
-            elif self.module.testing:
-                output = self.module.test_step(*inputs, **kwargs)
-            else:
-                output = self.module.validation_step(*inputs, **kwargs)
 
-        if not self._reducer_prepared_for_backwards and self.PREPARE_FOR_BACKWARDS:
-            self.reducer_prepare_for_backwards(output)
+class LightningDistributedModule(torch.nn.Module):
+
+    def __init__(self, pl_module: LightningModule):
+        """
+        Wraps the user's LightningModule and redirects the forward call to the appropriate
+        method, either ``training_step``, ``validation_step`` or ```test_step``.
+        This class is used in combination with :class:`~torch.nn.parallel.DistributedDataParallel` as
+        shown in the example.
 
-        if output is None:
-            warn_missing_output(f'{fx_called} returned None. Did you forget to return an output')
+        Example:
+
+            ddp_model = DistributedDataParallel(
+                module=LightningDistributedModule(lightning_module),
+                device_ids=[local_rank],
+                ...
+            )
+
+        Args:
+            pl_module: the model to wrap
+
+        """
+        super().__init__()
+        self.module = pl_module
+
+    def forward(self, *inputs, **kwargs):
+        if self.module.training:
```

Comment:
Any chance we could use the trainer state instead. If we user might change his model state by accident, we won't be calling the right function.

I am also thinking about people doing MC Dropout evaluation. `module.training` would be True if they don't set `training()`  properly.

Reply:
I thought about that for quite a while, but I think, we should rely on the module attribute, but not from self.module but from self.

Users would probably only change it based on their LightningModule and when we rely on self.training and self.testing (probably has to be added here then) we should be fine in that regard and did not tie it to close to the trainer, since we try to get rid of all the trainer references everywhere.

Reply:
I agree with both of you. The .training attribute is part of the nn.Module, however the .testing is set by the Trainer. This may not be so obvious, because there is no trace of training or testing attributes anywhere in the LightningModule. 
One attribute being part of nn.Module and one being part of LightningModule is strange and will not be so easy to debug. 

I'd like to keep it a strict refactor about the DDP class in this PR and not change the attributes yet.
Shall I create an issue so we can follow up on this?

Both of your suggestions seem reasonable to me. 
Justus's idea would require additional logic in the training loop to set the attributes on this wrapper.
Thomas's idea would basically be replacing self.module.training with self.module.trainer.training, right? We could also thing about adding read-only properties to the LightningModule as we did for other attributes that need to reference trainer.

Reply:
good, I would be just careful about adding too many new attributes to LightningModule :]